### PR TITLE
[SPARK-46460][Optimizer]The filter of partition including cast function may lead the partition pruning to disable

### PR DESCRIPTION
**Pull request by littlelittlewhite09**
### What changes were proposed in this pull request?
The issue is open on JIRA. Link:[SPARK-46460](https://issues.apache.org/jira/browse/SPARK-46460)

### Why are the changes needed?
The changes can avoid filter condition failure caused by cast function. But this optimization is only effective when cast string type to datetype.

### Does this PR introduce _any_ user-facing change?
no

### How was this patch tested?
UT

### Was this patch authored or co-authored using generative AI tooling?
no

### CODE REVIEW

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala [hunk 1]
```diff
@@ -2466,3 +2466,37 @@ object RemoveRepetitionFromGroupExpressions extends Rule[LogicalPlan] {
       }
   }
 }
+
+object EliminateCast extends Rule[LogicalPlan] {
```
Comment (HyukjinKwon) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala:2470]: I think you should improve `UnwrapCastInBinaryComparison`.

Comment (littlelittlewhite09) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala:2470]: ok, thanks. I will make a try

Comment (littlelittlewhite09) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala:2470]: @HyukjinKwon Hi, I checked UnwrapCastInBinaryComparison, and this rule may be applied on all filter conditions, including non-partitioned field, which may cause some bug if modify this rule. Maybe, modifying the `convertFilters` method in `HiveShim.scala` is a better way.

**Comment by littlelittlewhite09** [comment]
@HyukjinKwon Hi, Can you take few minutes to review code. Thanks a lot.

**Pull request closed without merging**